Show EU power values more readable

HybridDog opened the pull request:
Add the EU_string helper function
In comparison to pretty_num it uses SI prefixes, adds "EU" (e.g. kEU) and rounds the number for readability

Before:
![old](https://user-images.githubusercontent.com/3192173/39968789-0bde762c-56d3-11e8-8c78-d5bc5397d79d.jpg)

After:
![new](https://user-images.githubusercontent.com/3192173/39968792-10ef4d76-56d3-11e8-9778-e70955edbd6a.jpg)
(of course, the BB collected some power between taking the screenshots)

I've noticed that infotext is sometimes line-broken in minetest, which would puts the EU unit into the next line. Therefore I've changed the space to a non-breaking space character: ` `, are you fine with this?

numberZero commented on technic/helpers.lua:
```diff
@@ -21,6 +21,31 @@ function technic.pretty_num(num)
 
+local prefixes = {[-24] = "y", [-21] = "z", [-18] = "a", [-15] = "f",
+	[-12] = "p", [-9] = "n", [-6] = "µ", [0] = "", [-3] = "m", [3] = "k", [6] = "M",
+	[9] = "G", [12] = "T", [15] = "P", [18] = "E", [21] = "Z", [24] = "Y"}
```
Why do you step by 3 and not by 1?

HybridDog replied:
I thought this makes the code a bit easier to understand, I've changed it to 1.

numberZero commented on technic/helpers.lua:
```diff
@@ -21,6 +21,31 @@ function technic.pretty_num(num)
+local digits = 4  -- shouldn't be less than 4
+function technic.EU_string(num)
+	if num * 0 ~= num * 0 then
+		-- ±inf or ±nan
```
Ugh, Lua lacks `isfinite`...

HybridDog replied:
yes, the comment explains what is done here

numberZero commented on technic/helpers.lua:
```diff
@@ -21,6 +21,31 @@ function technic.pretty_num(num)
+	end
+	if not prefixes[b3] then
+		-- not likely going to happend
+		return tostring(num) .. " EU"
```
`string.format("%e EU", num)`

HybridDog replied:
I've used `"%.4g EU"` because l noticed that this is also used when num is 0.

numberZero commented on technic/helpers.lua:
```diff
@@ -21,6 +21,31 @@ function technic.pretty_num(num)
+		-- ±inf or ±nan
+		return tostring(num)
+	end
+	local b = math.floor(math.log(num) / math.log(10) +0.000001)
```
But Lua does have `log10`.
BTW, what if `num` is exactly zero?

HybridDog replied:
> But Lua does have log10.

Thanks, I didn't know this because l rarely use it…

> BTW, what if num is exactly zero?

See above.

numberZero commented on technic/helpers.lua:
```diff
@@ -21,6 +21,31 @@ function technic.pretty_num(num)
+		b = 0
+	else
+		b = b - (digits-1) * math.sign(b)
+	end
```
Some magic...

HybridDog replied:
For readability, it shows e.g. 4398 kEU instead of 4.298 MEU but to avoid showing 1001 mEU, as an edge case 1.001 EU is shown.

numberZero commented on technic/helpers.lua:
```diff
@@ -21,6 +21,31 @@ function technic.pretty_num(num)
+		return tostring(num) .. " EU"
+	end
+	num = math.floor(num / 10^b +.5) * 10^(b - b3)
+	return num .. " " .. prefixes[b3] .. "EU"
```
More magic...
Why not simply something like
```lua
b = math.floor(math.log10(num) / 3 + 0.001)
return string:format("%.2f %sEU", num / 1000^b, prefixes[b])
```
?

HybridDog replied:
This doesn't incorporate the edge case, see above.

HybridDog commented:
> why do you add new function instead of updating pretty_num?

When I wrote the new function I expected pretty_num to be used not only for power values but also for something different, but it turned out to being always used for power.
I thought I'd keep the new function name because pretty_num does not attach "EU" and the function name is therefore quite hazy.

HybridDog commented on technic/helpers.lua:
```diff
@@ -21,6 +21,30 @@ function technic.pretty_num(num)
+		return string.format("%.4g EU", num)
+	end
+
+	return string.format("%.4g %sEU", num / 10^(pref_i * 3), prefixes[pref_i])
```
What do you mean?
Specifying 4 digits is intended.

numberZero replied:
Sorry for confusing you. I understood that that’s how `%g` works, but I didn’t knew that before. It appears to be really handy)

HybridDog commented on technic/helpers.lua:
```diff
@@ -21,6 +21,30 @@ function technic.pretty_num(num)
+		pref_i = math.floor((b - 1) / 3)
+	else
+		-- as special case, avoid showing e.g. 1100 mEU instead of 1.1 EU
+		pref_i = math.floor(b / 3)
```
fixed

HybridDog commented:
> You could make the function append the Si prefix but not the unit tag, like 4.40 M. I would even prefer to see such function, so that it could be used for Power Units (PU, EU/s), for example (could be displayed by the supply converter, the power monitor, etc.)

I've changed the pretty_num function now, EU_string appends "EU" to the pretty_num function's result.

> So if you really prefer the heuristic you describe, I take back my objections

I prefer it because X.XXX numbers (X is digit) are avoided, XXXX is easier to understand in my opinion.

numberZero commented on technic/helpers.lua:
```diff
@@ -1,23 +1,33 @@
+	end
+	if not prefixes[pref_i] then
+		-- This happens for 0, nan, inf, very big values, etc.
+		return string.format("%.4g EU", num)
```
er, `EU` is still here?

numberZero commented on technic/helpers.lua:
```diff
@@ -1,23 +1,33 @@
 	end
-	digit_sep_esc = sep
+
+	return string.format("%.4g %sEU", num / 10^(pref_i * 3), prefixes[pref_i])
```
and here

numberZero commented on technic/helpers.lua:
```diff
@@ -1,23 +1,33 @@
-	return str
+-- used to display power values
+function technic.EU_string(num)
+	return technic.pretty_num(num) .. "EU"
```
...but added again here

numberZero commented on technic/doc/api.md:
```diff
@@ -11,9 +11,12 @@ switching station handles the network activity.
+* `technic.EU_string(num)`
+	* Converts num to a human-readable string (see pretty_num)
+	  and adds the `EU` unit
+	* Use this function when showing players power values
```
EU stands for *Energy* Unit, not *Power* Unit (which could be EU/s). Sadly technic uses EUs for both energy and power, which is incorrect.

HybridDog commented:
> Sadly technic uses EUs for both energy and power, which is incorrect.

EU/s is incorrect because the energy production speed depends on server lag.
#
I've fixed the EUEU mistake and added a constant digit count setting.

HybridDog commented:
`%.4g` emits zeros on the right, it shows e.g. "15" instead of "15.00"

numberZero commented:
Sorry, when I was talking about constant digit count, I didn’t know the exact `%g` behavior. It completely satisfies me.
So, I’m neutral on 78084a8, and :+1: on everything else. Sadly I’m only a contributor, and can’t merge anything...

SmallJoker commented:
This PR improves the readability a lot - I'm happy to see this change. However, could you please add unittests for `technic.pretty_num()` to ensure it outputs the strings as expected?

HybridDog commented:
I've added the unittests.
I've also disallowed showing `-0`.
